crates/bevy_render/src/mesh/mesh/mod.rs

```diff
@@ -426,7 +426,7 @@ impl MeshVertexAttribute {
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Hash)]
-pub struct MeshVertexAttributeId(usize);
+pub struct MeshVertexAttributeId(u64);
```

Comment:
Should we maybe add an implementation with an associated constant, something like:

```rust
impl MeshVertexAttributeId {
  const CUSTOM_ATTRIBUTE_ID_START: u64 = 0x100;
}
```

So that users know what range is safe, instead of a vague “close to zero” figure? We could even add something like this to `impl MeshVertexAttribute`:

```rust
pub const fn custom(name: &'static str, id: u64, format: VertexFormat) -> Self {
   Self::new(name, id + MeshVertexAttributeId::CUSTOM_ATTRIBUTE_ID_START, format)
}
```

(But this might be overkill)

Reply:
I was going to agree with something like that, but then I asked myself the question: why does Bevy reserve any range of attribute IDs for itself when the apparently isomorphic case of calls `HandleUntyped::weak_from_u64()` uses random values for all Bevy's built-in handles? So I looked and the comment explains that the attribute ID is also used as a sort key when ordering the data in the buffers. The implication is that this ordering is important for some reason (performance?), but it's not totally clear to me.

It the ordering is important then I think it would be better if the sort order could be specified independently of the ID. After all, what if you wanted to put your custom attribute before or in between the system ones? This would make the data structure in `Mesh` more complicated than the current `BTreeMap` of IDs though.

Either way, then the built-in vertex attributes could use random IDs too.